wgpu-hal/src/gles/device.rs

```diff
@@ -265,14 +267,60 @@ impl super::Device {
         unsafe { Self::compile_shader(gl, &output, naga_stage, stage.module.label.as_deref()) }
     }
 
-    unsafe fn create_pipeline<'a, I: Iterator<Item = ShaderStage<'a>>>(
+    unsafe fn create_pipeline<'a>(
         &self,
         gl: &glow::Context,
-        shaders: I,
+        shaders: ArrayVec<ShaderStage<'a>, 3>,
         layout: &super::PipelineLayout,
         #[cfg_attr(target_arch = "wasm32", allow(unused))] label: Option<&str>,
         multiview: Option<std::num::NonZeroU32>,
-    ) -> Result<super::PipelineInner, crate::PipelineError> {
+    ) -> Result<Arc<super::PipelineInner>, crate::PipelineError> {
+        let mut program_stages = ArrayVec::new();
+        let mut group_to_binding_to_slot = Vec::with_capacity(layout.group_infos.len());
+        for group in &*layout.group_infos {
+            group_to_binding_to_slot.push(group.binding_to_slot.clone());
+        }
+        for &(naga_stage, stage) in &shaders {
+            program_stages.push(super::ProgramStage {
+                naga_stage: naga_stage.to_owned(),
+                shader_id: stage.module.id,
+                entry_point: stage.entry_point.to_owned(),
+            });
+        }
+        let glsl_version = match self.shared.shading_language_version {
+            naga::back::glsl::Version::Embedded { version, .. } => version,
+            naga::back::glsl::Version::Desktop(_) => unreachable!(),
+        };
+        let mut guard = self.shared.program_cache.lock();
```

Comment:
This should be try_lock() - there is no situation where this is called where we aren't rlready behind the gl context lock

Reply:
We can't do anything with the program cache if the lock fails to acquire, so would you have it silently fail or panic on None?

Reply:
I would panic - it should be unreachable.

Reply:
We're basically using a lock as an atomic refcell

Reply:
But then isn't `try_lock().expect()` the same as `lock()`? Or do you specifically want an `unreachable!()` panic?

Reply:
Yeah - if we get it wrong with lock, we deadlock, if we get it wrong with try_lock, we panic which is much more discoverable.

Reply:
Ahh of course, I think I'm mixing up my locks. I'll switch it over!